aws/resource_aws_servicecatalog_product_test.go

```diff
@@ -0,0 +1,340 @@
+			return fmt.Errorf("Product ARN has not changed from %s, but it should have been re-created",
+				*describeProductOutput1.ProductViewDetail.ProductARN)
+		}
+		return nil
+	}
+}
+
+func testAccCheckAwsServiceCatalogProductNotRecreated(describeProductOutput1, describeProductOutput2 *servicecatalog.DescribeProductAsAdminOutput) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if *describeProductOutput1.ProductViewDetail.ProductARN != *describeProductOutput2.ProductViewDetail.ProductARN {
+			return fmt.Errorf("Product ARN has changed from %s to %s, but it should not have been re-created",
+				*describeProductOutput1.ProductViewDetail.ProductARN, *describeProductOutput2.ProductViewDetail.ProductARN)
+		}
+		return nil
+	}
+}
+
+func testAccCheckAwsServiceCatalogProductDisappears(describeProductOutput *servicecatalog.DescribeProductAsAdminOutput) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := testAccProvider.Meta().(*AWSClient).scconn
+		input := servicecatalog.DeleteProductInput{
+			Id: describeProductOutput.ProductViewDetail.ProductViewSummary.ProductId,
+		}
+		_, err := conn.DeleteProduct(&input)
+		if err != nil {
+			return fmt.Errorf("could not delete product: %s", err)
+		}
+		return nil
+	}
+}
+
+func testAccCheckAwsServiceCatalogProductDestroy(s *terraform.State) error {
+	count := 0
+	var rs *terraform.ResourceState
+	for _, rsi := range s.RootModule().Resources {
+		if rsi.Type != "aws_servicecatalog_product" {
+			continue
+		}
+		count++
+		rs = rsi
+	}
+	if count == 0 {
+		// disappears test
+		return nil
+	}
+	if count != 1 {
+		return fmt.Errorf("product count mismatch, found %d, expected 1", count)
+	}
+
+	conn := testAccProvider.Meta().(*AWSClient).scconn
+	input := servicecatalog.DescribeProductInput{}
+	input.Id = aws.String(rs.Primary.ID)
+
+	_, err := conn.DescribeProduct(&input)
+	if err == nil {
+		return fmt.Errorf("product still exists")
+	}
+	if !isAWSErr(err, servicecatalog.ErrCodeResourceNotFoundException, "") {
+		return err
+	}
+	// not found at AWS, as expected once destroyed
+	return nil
+}
+
+func testAccAWSServiceCatalogProductConfig_basic(resourceName, tag1, tag2, productSalt, bucketSalt, paSalt string) string {
+	thisResourceParts := strings.Split(resourceName, ".")
+	return fmt.Sprintf(`
+data "aws_region" "current" {}	
+
+resource "aws_s3_bucket" "bucket" {
+  bucket        = "bucket-%[6]s"
+  region        = data.aws_region.current.name
+  acl           = "private"
+  force_destroy = true
+}
+
+resource "aws_s3_bucket_object" "template1" {
+  bucket  = aws_s3_bucket.bucket.id
+  key     = "test_templates_for_terraform_sc_dev1.json"
+  content = <<EOF
+{
+  "AWSTemplateFormatVersion": "2010-09-09",
+  "Description": "Test CF teamplate for Service Catalog terraform dev",
+  "Resources": {
+    "Empty": {
+      "Type": "AWS::CloudFormation::WaitConditionHandle"
+    }
+  }
+}
+EOF
+}
+
+resource "%[1]s" "%[2]s" {
```

Comment:
try to avoid using this and align with other tests in the provider, this reduces readability

Reply:
okay - changed to hard-coded strings